File: LeetCode Problems/101 to 200/141. Linked List Cycle/141. Linked List Cycle.java
// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

class Solution {
    /*
        Floyd’s Cycle Detection Algorithm
        Time Complexity -> O(n)
        Space Complexoty -> O(1)

        Move the slow pointer by one position.
        Move the fast pointer by two positions.
        If both pointers meet at some point then a loop exists and if the fast pointer meets the end position then no loop exists.
    */
    public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) return true;
        }
        return false;
    }
}

File: LeetCode Problems/101 to 200/141. Linked List Cycle/141. Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        // Initialize two pointers: fast and slow, and assign them to the head of the list.
        ListNode *fast = head;
        ListNode *slow = head;

        // Traverse the list until fast and fast->next are not NULL.
        // Increment fast by 2 steps and slow by 1 step in each iteration.
        while (fast != nullptr && fast->next != nullptr) {
            fast = fast->next->next;
            slow = slow->next;

            // If fast and slow are at the same address, this means the linked list has a cycle.
            if (fast == slow) {
                return true;
            }
        }

        // If the traversal reaches NULL, this means there is no cycle.
        return false;
    }
};
